### ISSUE TITLE
Machine learning robustness analysis

### ISSUE THREAD

**Issue by gwaygenomics** [broadinstitute/cell-health#144]
One of the major questions in the review of our cell health paper involved questions of machine learning robustness. In #143 I add code, results, and figures for three separate analyses all assessing ML robustness.

# Analyses

1. Cell Line Holdout - Predict cell health readouts from one cell line completely held out of training
2. Sample Titration - Systematically remove samples and assess test set performance degradation
3. Feature Removal - Remove Cell Painting features and assess test set performance changes

## 1. Cell line holdout analysis

![cell_line_holdout_modz](https://user-images.githubusercontent.com/7340421/98130516-55467b00-1e88-11eb-8dbd-d14fb63767a9.png)

**Figure legend. Predicting cell health readouts in held out cell lines.** We retrained a new model for each cell health variable using profiles from only 2 of 3 cell lines. We applied the models to predict the readouts from the third cell line. We show these predictions for each cell line when included in the held out set.

Even across cell lines and with a much smaller training set set (357 * 2/3 = 238), many models still detect signal more than a negative control. Many phenotypes can be predicted (e.g. viability, s phase, mitosis, death) in cell lines the model has never seen before, but many other phenotypes cannot be predicted. Not surprisingly, the shape phenotypes could not be predicted reliably in held out cell lines.

## 2. Sample titration

![sample_titration_modz](https://user-images.githubusercontent.com/7340421/97022722-74f1b100-1522-11eb-9c54-f80d8e7286e1.png)

**Figure legend. The impact of removing samples from model training.** We show the test set R squared of a sample titration analysis in which we randomly remove up to 275 samples from the training procedure. We perform 10 different iterations of the sample titration analysis (so different samples are dropped 10 different times across the full titration range). We present the mean value of all iterations per titration amount.

## 3. Systematic feature group removal

![feature_removal_modz](https://user-images.githubusercontent.com/7340421/95129413-f971c080-0728-11eb-8391-b5cd98dfe5a2.png)

**Figure legend. Removing entire feature groups from training cell health models** We show the performance difference between the original cell health models and new models trained with the corresponding (a) feature group, (b) channel, and (c) compartment removed before model training. The `n` indicates the number of features removed and each point is an individual cell health model. (d) The impact of feature removal on individual cell health models.

As expected, systematically removing features does not have much of an impact on average for all feature classes. There are a handful of outliers, however. These outliers are driven mostly by a small minority of cell health models. It is also interesting that many of the models actually _increased_ in performance (negative values on x axis). For example, the nucleus roundess phenotype as predicted better when any set of features were removed.

**Comment by AnneCarpenter** [comment]
My conclusion from 2. Sample titration is that there is plenty of room on that curve moving left-ward - i.e. adding more data certainly seems like it would continue to improve the models. We assumed this must be the case given how small our training sets are, but it's nice to see it confirmed.

**Comment by gwaygenomics** [comment]
Copying Shantanu's comment from the response document, in response to the question about how much signal loss exists when performing the cell line holdout analysis.

---

Grouping all 70 models together, per cell line, is not informative(below)
Alternative: show %-decrease in performance per model per cell line, and plot that as a distribution.

https://github.com/broadinstitute/cell-health/blob/cf7209965e28c13f79ea2ca831b0968f4aae02ed/6.ml-robustness/results/cell_line_holdout_robustness_results_modz.tsv.gz
x <- read_tsv("~/Downloads/cell_line_holdout_robustness_results_modz.tsv.gz")
x %>% filter(metric == "r_two" & data_fit == "test" & shuffle == "shuffle_false")   %>% ggplot(aes(cell_line, value)) + geom_boxplot()

![Picture1](https://user-images.githubusercontent.com/7340421/98130764-9a6aad00-1e88-11eb-8c05-815617fdcc9e.png)

x %>% filter(metric == "r_two" & data_fit == "test" & shuffle == "shuffle_false")  %>% group_by(cell_line) %>% summarize_each(c("value"), funs = list(mean = mean, sd = sd))

```
# A tibble: 3 x 3
  cell_line     mean    sd
  <chr>        <dbl> <dbl>
1 A549       0.00297 0.280
2 ES2        0.0492  0.248
3 HCC44     -0.0370  0.518
```

**Comment by AnneCarpenter** [comment]
@shntnu What is your conclusion from that analysis?
